File: src/domain/Blind.cpp
#include "Blind.h"
#include <algorithm>
#include <cmath>

namespace Blinds {
    namespace Domain {

        Blind::Blind(Interfaces::IMotor& motor,
                     Interfaces::ILimitSwitch* topLimit,
                     Interfaces::ILimitSwitch* bottomLimit,
                     Interfaces::IStorage& storage,
                     BlindConfig config)
            : _motor(motor),
              _topLimit(topLimit),
              _bottomLimit(bottomLimit),
              _storage(storage),
              _config(config),
              _currentState(BlindState::UNKNOWN)
        {
        }

        void Blind::initialize() {
            _motor.setSpeedInHz(_config.defaultSpeed);
            _motor.setAcceleration(_config.acceleration);
            _motor.enable(true);

            int32_t savedPos = _storage.loadPosition();
            // Simple validation of saved position
            if (savedPos >= 0 && savedPos <= _config.maxTravelSteps) {
                 _motor.setCurrentPosition(savedPos);
                 setState(BlindState::IDLE);
            } else {
                setState(BlindState::UNKNOWN);
                // If we have limits, we should probably home.
                // But we wait for explicit home command or startHoming() call.
            }
        }

        void Blind::update() {
            checkLimits();

            // State Machine for Homing
            if (_currentState == BlindState::HOMING) {
                if (_homingState == HomingState::MOVING_TO_BOTTOM) {
                    if (_bottomLimit && _bottomLimit->isTriggered()) {
                        _motor.stop();
                        _motor.setCurrentPosition(_config.maxTravelSteps);
                        _homingState = HomingState::RETRACTING;
                        // Retract slightly to clear switch
                        _motor.move(-100);
                    } else if (!_motor.isRunning()) {
                        // We stopped but didn't hit limit? Maybe stalled or manually stopped.
                         // For now, if we reach here and motor stopped, fail?
                         // Or if we commanded a huge move and it finished (unlikely if limit switch works)
                    }
                } else if (_homingState == HomingState::RETRACTING) {
                    if (!_motor.isRunning()) {
                        _homingState = HomingState::DONE;
                        setState(BlindState::IDLE);
                        _storage.savePosition(_config.maxTravelSteps);
                    }
                }
                return;
            }

            // Normal operation monitoring
            if (_motor.isRunning()) {
                // Direction detection?
                // For now, just rely on state set by commands, or infer from velocity if available.
                // But we need to know if we are moving UP or DOWN to set state correctly if we just check isRunning.
                // However, commands set the state.
            } else {
                if (_currentState == BlindState::MOVING_UP || _currentState == BlindState::MOVING_DOWN) {
                    setState(BlindState::IDLE);
                    _storage.savePosition(_motor.getCurrentPosition());
                }
            }
        }

        void Blind::checkLimits() {
            if (_homingState != HomingState::NONE) return; // Homing handles its own limits

            if (_topLimit && _topLimit->isTriggered()) {
                // If moving up, stop
                // We don't easily know current direction from IMotor without polling velocity,
                // but we can assume if we are at limit, we should hard stop and set pos.
                if (_currentState == BlindState::MOVING_UP) {
                    _motor.emergencyStop();
                    _motor.setCurrentPosition(0);
                    setState(BlindState::IDLE);
                }
            }

            if (_bottomLimit && _bottomLimit->isTriggered()) {
                if (_currentState == BlindState::MOVING_DOWN) {
                    _motor.emergencyStop();
                    _motor.setCurrentPosition(_config.maxTravelSteps);
                    setState(BlindState::IDLE);
                }
            }
        }

        void Blind::moveUp() {
            if (_currentState == BlindState::HOMING || _currentState == BlindState::FAULT) return;

            // Soft limit
            if (_motor.getCurrentPosition() <= 0 && !_config.hasLimitSwitches) {
                return;
            }

            _motor.moveTo(0);
            setState(BlindState::MOVING_UP);
        }

        void Blind::moveDown() {
            if (_currentState == BlindState::HOMING || _currentState == BlindState::FAULT) return;

            // Soft limit
            if (_motor.getCurrentPosition() >= _config.maxTravelSteps && !_config.hasLimitSwitches) {
                return;
            }

            _motor.moveTo(_config.maxTravelSteps);
            setState(BlindState::MOVING_DOWN);
        }

        void Blind::moveToPercent(int percent) {
            if (_currentState == BlindState::HOMING || _currentState == BlindState::FAULT) return;

            percent = std::max(0, std::min(100, percent));
            int32_t target = percentToSteps(percent);

            int32_t current = _motor.getCurrentPosition();
            if (target < current) {
                setState(BlindState::MOVING_UP);
            } else if (target > current) {
                setState(BlindState::MOVING_DOWN);
            } else {
                return; // Already there
            }

            _motor.moveTo(target);
        }

        void Blind::stop() {
            _motor.stop();
            setState(BlindState::IDLE); // update() will catch the actual stop
        }

        void Blind::startHoming() {
            if (!_config.hasLimitSwitches || !_bottomLimit) return; // Cannot home without bottom limit

            setState(BlindState::HOMING);
            _homingState = HomingState::MOVING_TO_BOTTOM;
            _motor.setSpeedInHz(_config.homingSpeed);

            // Move huge distance down
            _motor.move(1000000);
        }

        int Blind::getPositionPercent() {
            return stepsToPercent(_motor.getCurrentPosition());
        }

        int32_t Blind::getCurrentSteps() {
            return _motor.getCurrentPosition();
        }

        BlindState Blind::getState() {
            return _currentState;
        }

        void Blind::setState(BlindState newState) {
            _currentState = newState;
        }

        int32_t Blind::percentToSteps(int percent) {
            return (int32_t)((double)percent / 100.0 * _config.maxTravelSteps);
        }

        int Blind::stepsToPercent(int32_t steps) {
            if (_config.maxTravelSteps == 0) return 0;
            int pc = (int)((double)steps / _config.maxTravelSteps * 100.0);
            return std::max(0, std::min(100, pc));
        }

        std::string Blind::getStateString() {
            switch (_currentState) {
                case BlindState::UNKNOWN: return "UNKNOWN";
                case BlindState::IDLE: return "IDLE";
                case BlindState::MOVING_UP: return "MOVING_UP";
                case BlindState::MOVING_DOWN: return "MOVING_DOWN";
                case BlindState::HOMING: return "HOMING";
                case BlindState::FAULT: return "FAULT";
                default: return "UNKNOWN";
            }
        }
    }
}

File: src/domain/blind.py
from typing import Optional
from ..interfaces.hardware import IMotor, ILimitSwitch, IStorage
from .config import BlindConfig
from .state import BlindState, HomingState

class Blind:
    def __init__(self, motor: IMotor, storage: IStorage, config: BlindConfig,
                 top_limit: Optional[ILimitSwitch] = None,
                 bottom_limit: Optional[ILimitSwitch] = None):
        self._motor = motor
        self._storage = storage
        self._config = config
        self._top_limit = top_limit
        self._bottom_limit = bottom_limit

        self._current_state = BlindState.UNKNOWN
        self._homing_state = HomingState.NONE

    @property
    def config(self) -> BlindConfig:
        return self._config

    def initialize(self):
        self._motor.set_speed(self._config.default_speed)
        self._motor.set_acceleration(self._config.acceleration)
        self._motor.enable(True)

        saved_pos = self._storage.load_position()
        if saved_pos is not None and 0 <= saved_pos <= self._config.max_travel_steps:
            self._motor.set_current_position(saved_pos)
            self.set_state(BlindState.IDLE)
        else:
            self.set_state(BlindState.UNKNOWN)

    def update(self):
        self.check_limits()

        # Homing State Machine
        if self._current_state == BlindState.HOMING:
            if self._homing_state == HomingState.MOVING_TO_BOTTOM:
                if self._bottom_limit and self._bottom_limit.is_triggered():
                    self._motor.stop()
                    self._motor.set_current_position(self._config.max_travel_steps)
                    self._homing_state = HomingState.RETRACTING
                    self._motor.move(-100) # Retract
            elif self._homing_state == HomingState.RETRACTING:
                if not self._motor.is_running():
                    self._homing_state = HomingState.DONE
                    self.set_state(BlindState.IDLE)
                    self._storage.save_position(self._config.max_travel_steps)
            return

        # Regular monitoring
        if not self._motor.is_running():
            if self._current_state in [BlindState.MOVING_UP, BlindState.MOVING_DOWN]:
                self.set_state(BlindState.IDLE)
                self._storage.save_position(self._motor.get_current_position())

    def check_limits(self):
        if self._homing_state != HomingState.NONE:
            return

        if self._top_limit and self._top_limit.is_triggered():
            if self._current_state == BlindState.MOVING_UP:
                self._motor.emergency_stop()
                self._motor.set_current_position(0)
                self.set_state(BlindState.IDLE)

        if self._bottom_limit and self._bottom_limit.is_triggered():
            if self._current_state == BlindState.MOVING_DOWN:
                self._motor.emergency_stop()
                self._motor.set_current_position(self._config.max_travel_steps)
                self.set_state(BlindState.IDLE)

    def move_up(self):
        if self._current_state in [BlindState.HOMING, BlindState.FAULT]:
            return

        if self._motor.get_current_position() <= 0 and not self._config.has_limit_switches:
            return

        self._motor.move_to(0)
        self.set_state(BlindState.MOVING_UP)

    def move_down(self):
        if self._current_state in [BlindState.HOMING, BlindState.FAULT]:
            return

        if self._motor.get_current_position() >= self._config.max_travel_steps and not self._config.has_limit_switches:
            return

        self._motor.move_to(self._config.max_travel_steps)
        self.set_state(BlindState.MOVING_DOWN)

    def move_to_percent(self, percent: int):
        if self._current_state in [BlindState.HOMING, BlindState.FAULT]:
            return

        percent = max(0, min(100, percent))
        target_steps = int((percent / 100.0) * self._config.max_travel_steps)

        current_steps = self._motor.get_current_position()
        if target_steps < current_steps:
            self.set_state(BlindState.MOVING_UP)
        elif target_steps > current_steps:
            self.set_state(BlindState.MOVING_DOWN)
        else:
            return

        self._motor.move_to(target_steps)

    def stop(self):
        self._motor.stop()
        self.set_state(BlindState.IDLE)

    def start_homing(self):
        if not self._config.has_limit_switches or not self._bottom_limit:
            return

        self.set_state(BlindState.HOMING)
        self._homing_state = HomingState.MOVING_TO_BOTTOM
        self._motor.set_speed(self._config.homing_speed)
        self._motor.move(100000) # Move down indefinitely

    def get_position_percent(self) -> int:
        if self._config.max_travel_steps == 0:
            return 0
        steps = self._motor.get_current_position()
        pc = (steps / self._config.max_travel_steps) * 100.0
        return max(0, min(100, int(pc)))

    def get_current_steps(self) -> int:
        return self._motor.get_current_position()

    def get_state(self) -> BlindState:
        return self._current_state

    def set_state(self, new_state: BlindState):
        self._current_state = new_state
